### PR TITLE
Display advanced block propagation technology (XThin/Graphene) statistics in the Debug Window UI

### PR DESCRIPTION
**Pull request by Justaphf**
Now that graphene is merged into `dev`, I wanted a quick way to monitor which block propagation technologies are being actively used directly in the QT UI.  This is just displaying the number of blocks sent/received by XThin/Graphene. Does not currently include any of the other more detailed information available via `getnetworkinfo`.  Stats are automatically updated in the UI using one of the pre-existing timers for the debug window.

### CODE REVIEW

**Comment by Justaphf** [comment]
Putting in WIP, noticed some layout issues after resizing the dialog.  Likely a result of me being new to Qt Creator for doing wysiwyg layouts.

**Review by sickpig — COMMENTED** [review]
@Justaphf rebase needed (I know this is still WIP but still)

**Comment by Justaphf** [comment]
Taking out of WIP.  Everything is working now in terms of displaying what, IMO, were the most prominent summary statistics for both XThin and Graphene on the debug dialog.  This is a subset of the full statistics available via the `getnetworkinfo` RPC call.

Looking for feedback on the UI layout and category/label naming.  I tried to combine everything in a compact way that doesn't require the UI dialog size to be too large to see everything.  I grouped by technology type, then statistic time-frame, but maybe it makes more sense to group by time-frame then by technology type.  I tried it both ways and see benefits for both.

NOTE: As part of this, I updated the UI layout via Qt Creator.  All fields on the first tab are in a 3-column grid layout.  I changed it such that title rows (bold text) now span all 3 columns, label-value rows are defined as label 1 column, value spans 2 columns.  This caused a lot of churn in the debugwindow.ui file.  Not sure if making these changes caused some unintended side-effect with the UI layout, but seems to be OK on a Windows 10 test system.

EDIT: Removed side-bar comment.

File: src/thinblock.cpp
Comment (awemany) [inline src/thinblock.cpp]: Nit: Maybe `block(s)` instead of `block` to properly deal with the singular case :)

File: src/graphene.cpp
Comment (awemany) [inline src/graphene.cpp]: Typo in the function name? 'comput'

File: src/graphene.cpp
Comment (awemany) [inline src/graphene.cpp]: nit: IMO `for (auto|<type> m : ...) {...}` is more readable

File: src/graphene.cpp
Comment (awemany) [inline src/graphene.cpp]: Can you further DRY that with the inbound function above? Seems like they share a lot of code still...

File: src/thinblock.cpp
Comment (awemany) [inline src/thinblock.cpp]: Hmm, this is an almost repeat of `compute24hInboundCompressionInternal()`, isn't it?

**Review by awemany — COMMENTED** [review]
Are the newly computed stats also available via RPC?

File: src/graphene.cpp
Comment (Justaphf) [reply src/graphene.cpp]: Will fix.  I also noticed some spelling/typos in some of the comments as well.

File: src/graphene.cpp
Comment (Justaphf) [reply src/graphene.cpp]: I hadn't changed the original algorithms beyond moving them into helper methods.   I can refactor this and any other locations where I touched the code.

File: src/graphene.cpp
Comment (Justaphf) [reply src/graphene.cpp]: Will do.

File: src/thinblock.cpp
Comment (Justaphf) [reply src/thinblock.cpp]: Will make the same DRY updates as requested for graphene.

File: src/thinblock.cpp
Comment (Justaphf) [reply src/thinblock.cpp]: I think that since there are always 2 numbers listed now it is grammatically correct to always use the plural.

**Comment by Justaphf** [comment]
> Are the newly computed stats also available via RPC?

The stats aren't new, I just extracted the algorithms from the existing methods that are called to construct the RPC outputs used by `getnetworkinfo` and put them in a helper struct to easily pass to the UI for formatting and display.

**Comment by ptschip** [comment]
@Justaphf been running this today, the only thing i'd like to bikeshed about is the compression ratios having 2 decimal places rather than one in the getnetworkinfo.  I think the 2 just adds to visual noise and IMO we should be consistent with getnetworkinfo.  Other than that I think it all looks great!

**Comment by Justaphf** [comment]
@ptschip I'm working on the comments from @awemany presently and will make the precision consistent for the compression percentages as well.

I also found a bug in the compression statistics computations for both XThin and Graphene that I'm fixing.  Looks like we expire the block maps at the time of the computation, but don't expire the bloom filter/mempool info maps, which are also used in the computation.  This could potentially cause extra values to remain in the totals for bloom filter/mempool info for a short window, producing slightly lower compression percentages.

Edit: typo

File: src/graphene.cpp
Comment (Justaphf) [reply src/graphene.cpp]: Fixed.

File: src/graphene.cpp
Comment (Justaphf) [reply src/graphene.cpp]: Fixed for all old style for loops in thinblock.cpp and graphene.cpp related to stats computations.

File: src/graphene.cpp
Comment (Justaphf) [reply src/graphene.cpp]: Fixed

File: src/thinblock.cpp
Comment (Justaphf) [reply src/thinblock.cpp]: Fixed

**Comment by Justaphf** [comment]
Fixed review comments.

Additionally corrected the stats computations to ensure that all stats maps used in computations have called `expireStats` prior to computation to ensure the data being computed correctly spans the specified 24-hour period.  Two potential errors were caught here.  Some of the computations use values from 2 maps, but only called `expireStats` on one of the maps.  Other computations didn't call `expireStats` at all so could exceed the 24-hour window.

NOTE: Several of the refactors I left as separate commits to make it easier to review.  Can squash up once approved for merge if so desired.

File: src/thinblock.cpp
Comment (awemany) [reply src/thinblock.cpp]: Indeed!

File: src/graphene.cpp
Comment (awemany) [reply src/graphene.cpp]: Great!

File: src/thinblock.cpp
Comment (ptschip) [inline src/thinblock.cpp]: why do we have to track if thinblocks is enabled by setting stats.fEnabled, when we can always get the current value from IsThinBlocksEnabled()?

File: src/thinblock.cpp
Comment (awemany) [reply src/thinblock.cpp]: Great, other change as well!

File: src/thinblock.cpp
Comment (Justaphf) [reply src/thinblock.cpp]: You are correct.  I will remove and directly use the Is*Enabled() calls for thin blocks and graphene in the UI display logic.

**Comment by Justaphf** [comment]
Correct @ptschip last comment and rebased.

**Review by ptschip — APPROVED** [review]
looks good, thanks for doing this...

**Pull request merged**
